cmd/main.go

```diff
@@ -86,16 +90,58 @@ func main() {
 	// Register routes - no need for project and task handlers as they're handled via RPC
 	routes.RegisterRoutes(app, instanceHandler, userHandler, rpcHandler)
 
-	// Start server
-	port := os.Getenv("SERVER_PORT")
-	if port == "" {
-		port = "8080"
+	// Create context that listens for the interrupt signal from the OS.
+	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer stop()
+
+	// Create a WaitGroup to wait for goroutines to finish
+	var wg sync.WaitGroup
+
+	// Launch worker with the cancellable context and WaitGroup
+	wg.Add(1) // Increment counter before launching goroutine
+	go services.LaunchWorker(ctx, &wg, taskService)
+
+	// Start server in a goroutine so that it doesn't block.
+	var errChan = make(chan error)
+	go func() {
+		port := os.Getenv("SERVER_PORT")
+		if port == "" {
+			port = "8080"
+		}
+		log.Info("Server starting on :" + port)
+		if err := app.Listen(":" + port); err != nil {
+			// Send error to the channel
+			errChan <- err
+		}
+	}()
+
+	// Listen for the interrupt signal or the server error.
+	select {
+	case err := <-errChan:
+		log.Errorf("Server failed to start: %v", err)
+	case <-ctx.Done():
+		log.Info("Received interrupt signal, shutting down gracefully")
 	}
 
-	log.Info("Server starting on :" + port)
-	if err := app.Listen(":" + port); err != nil {
-		log.Fatalf("Failed to start server: %v", err)
+	// Restore default behavior on the interrupt signal and notify user of shutdown.
+	stop()
+	log.Info("Shutting down gracefully, press Ctrl+C again to force")
+
+	// Perform application shutdown with a timeout.
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second) // 5 second timeout for shutdown
```

Comment:
[nit]: that 5 second could be extracted as a defined constant

Reply:
yup, when someone is bothered enough by it they can do it :-)